World Tool

meniku opened the pull request:
This is the first of two planned tools for modifying world-files within tiled. 

It allows you to:
1. Create new worlds in the "Map" menu
2. Save modified worlds in the "Map" menu or when you close tiled
3. add the current map to a loaded world via the World-Tool (hotkey: N) and then right-clicking in the scene
4. remove the current map from a loaded world via the World-Tool (hotkey: N) and then right-clicking in the scene
5. add other maps to a loaded world via the World-Tool (hotkey: N) and then right-clicking in the scene
6. Move the current map within the  World-Tool (hotkey: N) and then dragging it with your mouse 
7. Move the current map within the  World-Tool (hotkey: N) and then using Arrow-Keys (shift modifier will make movement steps bigger)

I think that's about it for now.
Functionalities 3-5 are basically just placeholders for the second tool I want to create at some point: a panel containing a list of all available maps which you can drag into the world or drag out of it. However right now I don't feel the need for myself so it could take some time until I get to it.

The tool doesn't currently allow modifying maps which make use of patterns, also it doesn't preserve any other custom properties you place into the world file, so use with care and backup the world file before using the world tools.

bjorn commented:
Hey @meniku, sorry for still not getting around to this! In the past weeks I really wanted to finish the initial support for projects and sessions, and this week I've been doing a lot of reviewing but didn't have time for this one yet. Next week it's Christmas so it may still take a bit to get around to it, but it's on the list.

bjorn commented:
I've rebased these commits onto my latest `master`, so that it now includes the macOS workflow files.

Please note that in general it's not recommended to open pull requests from the `master` branch. It's better to set up a dedicated branch in your fork for the pull request, to avoid pushing commits unrelated to the pull request.

meniku commented:
hey thanks for reviewing and eve considering this to be included in Tiled. I'll give my best to tackle the issues and improvements you mentioned.

meniku commented on src/tiled/mainwindow.cpp:
```diff
@@ -569,6 +571,53 @@ MainWindow::MainWindow(QWidget *parent, Qt::WindowFlags flags)
+            mUi->menuUnloadWorld->setEnabled(!worldFiles.isEmpty());
+        }
+    });
+    connect(mUi->menuSaveWorld, &QMenu::aboutToShow, this, [this] {
```
I think you're right, although here it doesn't appear to be disabled, regardless of whether a world is loaded or not. Do you think it's a big deal if we just keep the menu itself active even when there's no items in it?

bjorn replied:
I think disabling when there are no worlds to save is good, so a possible solution would be to not wait with populating the submenu until it is about to show. However, if we decide to move towards a world editor instead of a tool then there is no point in fixing this behavior since this menu item will go away.

meniku replied:
Your call, I'll not touch it until you decide. ( same for the usability improvements )

meniku commented on src/tiled/documentmanager.cpp:
```diff
@@ -905,8 +911,6 @@ void DocumentManager::currentIndexChanged()
 
     if (document) {
         editor = mEditorForType.value(document->type());
-        mUndoGroup->setActiveStack(document->undoStack());
```
I've fixed this by setting the active undoStack when the Tileset editor activates

meniku commented:
> In any case, I'll try to help out with this in order to include the feature in Tiled 1.4. I expect to have some time for this next week. It would be great if you could try to address as much of the feedback in the meantime.

Yeah, I've fixed some of the low hanging fruits. For the remaining stuff (mainly all those usability improvements) I would only go for them if you decide to keep it as a tool

bjorn commented:
> For the remaining stuff (mainly all those usability improvements) I would only go for them if you decide to keep it as a tool (else that work might be lost anyway in the end)

I fully understand. Do you have any opinion on the choice between tool or editor?

meniku commented:
> Do you have any opinion on the choice between tool or editor?
From a user's perspective I personally don't care. However from the perspective as being the one who implements it, I prefer to keep the tool and fix the usability issues you mentioned. I know there's quite some code changes to the existing code that wouldn't be necessary if it would be a separate editor, so I think you have to be the one deciding it.

bjorn commented:
@meniku Did you miss some of my feedback because they appear collapsed here on GitHub by default? There were a few more comments that should be easy to address.

I'll try to move it to an editor as soon as I find time for that, and decide whether it's better depending on how it turns out.

meniku commented:
> @meniku Did you miss some of my feedback because they appear collapsed here on GitHub by default? There were a few more comments that should be easy to address.
> 
> I'll try to move it to an editor as soon as I find time for that, and decide whether it's better depending on how it turns out.

ah possible, I'm gonna check them.

meniku commented on src/libtiled/worldmanager.cpp:
```diff
@@ -258,7 +474,9 @@ bool World::containsMap(const QString &fileName) const
 QRect World::mapRect(const QString &fileName) const
 {
-    for (const World::MapEntry &mapEntry : maps) {
+    const QVector<World::MapEntry> all(allMaps());
```
probably copy pasted it from somewhere else

meniku commented on src/tiled/worlddocument.h:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Represents an editable world document.
+ */
+class WorldDocument : public QObject
```
It's quite a while ago now. Probably I didn't feel the need to, because I didn't need anything of the base functionality. Also world file handling was in some separate class (worldmanager) already.

I see there's methods for saving etc in the base class. Maybe we should move things over, but that could be quite some work. Have to investigate this further when I have some time.

meniku commented:
Hi Bjorn,
I worked a bit on the usability feedback you gave me.

> When you move the current map, the result is unintuitive because all other maps appear to shift instead. That of course happens because indeed the other map items are repositioned relative to the current one, to keep the current one at the origin. Maybe that's something that does not need to happen.

I "fixed" this by undoing the camera movement after the map is moved. There's probably better ways, but it does the trick for me.

> When right clicking, a menu allows removing a map from the world. But it always removes the current map, whereas it should really remove the map that was clicked. It should be possible to implement this based on the mouse position.

I changed it as you suggested. I also don't allow you to remove the current map from the world anymore to remove confusion with the undo stack.

> Only the current map can be dragged. It should be possible to drag around any map.

I changed it as you suggested. Any map can now be dragged. Howevered maps can be moved with the arrow keys.

meniku commented:
> I think it's unexpected and easy to miss that the arrow keys will move the map in the world. Might be fixed by adding a clear selection overlay that emphasizes the world editing mode.

I've pushed a change to address this feedback. I'll see if I can get the icons a bit improved. Also I've a bit trouble positioning the context menu properly for adding the current map to a loaded world

bjorn commented:
So today I've done some more fixes and cleanups. There are a few remaining topics:

* I'm not sure about the addition of the directional move buttons on the tool bar. I understand you want to make those actions discoverable, but it seems overly cluttered to me. In addition, the generic "Left", "Right" etc. shortcuts interfere with changing the selected items in the various views.

* I'd still like to see the map move and the selection rectangle only used to mark its old location, instead of moving the selection rectangle.

* `AbstractWorldTool::addAnotherMapToWorld` will only ask the user for `.tmx` files, whereas it should support any map file.

bjorn commented:
I think this pull request is pretty much ready for merging at this point. The main thing still missing is the _documentation_... Would you be up for writing a first draft, @meniku? :-)

Also please let me know if there's anything left on your mind regarding merging this.

meniku commented:
> I think this pull request is pretty much ready for merging at this point. The main thing still missing is the _documentation_... Would you be up for writing a first draft, @meniku? :-)
> 
> Also please let me know if there's anything left on your mind regarding merging this.

I've added some preliminary documentation, might need some formatting
